### PR TITLE
Few more IDE scenarios for top-level statements

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/44272 (RemoveUnusedLocalFunction)
Fixes https://github.com/dotnet/roslyn/issues/44273 (RemoveUnusedVariable)
Fixes ConvertSwitchStatementToExpression

### CODE REVIEW

**Comment by jcouv** [comment]
Tagging @CyrusNajmabadi @sharwell @AlekseyTs

File: src/Analyzers/CSharp/CodeFixes/ConvertSwitchStatementToExpression/ConvertSwitchStatementToExpressionCodeFixProvider.cs [hunk 1]
```diff
@@ -81,7 +81,8 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
                     declaratorToRemoveTypeOpt = semanticModel.GetDeclaredSymbol(declaratorToRemoveNodeOpt).GetSymbolType();
                 }
 
-                var switchStatement = (SwitchStatementSyntax)switchLocation.FindNode(cancellationToken);
+                var switchNode = switchLocation.FindNode(getInnermostNodeForTie: true, cancellationToken);
+                var switchStatement = (SwitchStatementSyntax)(switchNode);
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/ConvertSwitchStatementToExpression/ConvertSwitchStatementToExpressionCodeFixProvider.cs:85]:
```suggestion
                var switchStatement = (SwitchStatementSyntax)switchLocation.FindNode(getInnermostNodeForTie: true, cancellationToken);
```

File: src/Analyzers/CSharp/CodeFixes/ConvertSwitchStatementToExpression/ConvertSwitchStatementToExpressionCodeFixProvider.cs [hunk 2]
```diff
@@ -97,8 +98,12 @@ public override Task RegisterCodeFixesAsync(CodeFixContext context)
...
                 {
                     // Already morphed into the top-level switch expression.
-                    var nextStatement = switchStatement.GetNextStatement();
+                    SyntaxNode nextStatement = switchStatement.GetNextStatement();
                     Debug.Assert(nextStatement.IsKind(SyntaxKind.ThrowStatement, SyntaxKind.ReturnStatement));
+                    if (nextStatement.IsParentKind(SyntaxKind.GlobalStatement))
+                    {
+                        nextStatement = nextStatement.Parent;
+                    }
                     editor.RemoveNode(nextStatement);
```
Comment (CyrusNajmabadi) [inline src/Analyzers/CSharp/CodeFixes/ConvertSwitchStatementToExpression/ConvertSwitchStatementToExpressionCodeFixProvider.cs:107]:
```suggestion
                    editor.RemoveNode(nextStatement.IsParentKind(SyntaxKind.GlobalStatement) ? nextStatement.Parent : nextStatement);
```

(can keep teh var above as well).

File: src/Features/CSharp/Portable/RemoveUnusedLocalFunction/CSharpRemoveUnusedLocalFunctionCodeFixProvider.cs [hunk 3]
```diff
@@ -58,7 +58,12 @@ protected override Task FixAllAsync(Document document, ImmutableArray<Diagnostic
 
             foreach (var localFunction in localFunctions)
             {
-                editor.RemoveNode(localFunction);
+                SyntaxNode nodeToRemove = localFunction;
+                if (localFunction.Parent.Kind() == SyntaxKind.GlobalStatement)
+                {
+                    nodeToRemove = localFunction.Parent;
+                }
+                editor.RemoveNode(nodeToRemove);
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/RemoveUnusedLocalFunction/CSharpRemoveUnusedLocalFunctionCodeFixProvider.cs:66]: same feedback.  

also, i'm really hating this pattern :-/

File: src/Features/CSharp/Portable/RemoveUnusedLocalFunction/CSharpRemoveUnusedLocalFunctionCodeFixProvider.cs [hunk 4]
```diff
@@ -58,7 +58,12 @@ protected override Task FixAllAsync(Document document, ImmutableArray<Diagnostic
 
             foreach (var localFunction in localFunctions)
             {
-                editor.RemoveNode(localFunction);
+                SyntaxNode nodeToRemove = localFunction;
+                if (localFunction.Parent.Kind() == SyntaxKind.GlobalStatement)
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/RemoveUnusedLocalFunction/CSharpRemoveUnusedLocalFunctionCodeFixProvider.cs:62]:
```suggestion
                if (localFunction.IsParentKind(SyntaxKind.GlobalStatement))
```

File: src/Features/CSharp/Portable/RemoveUnusedVariable/CSharpRemoveUnusedVariableCodeFixProvider.cs [hunk 5]
```diff
@@ -59,6 +59,10 @@ protected override void RemoveOrReplaceNode(SyntaxEditor editor, SyntaxNode node
                     editor.ReplaceNode(node, ((AssignmentExpressionSyntax)node).Right);
                     return;
                 default:
+                    if (node.Parent.Kind() == SyntaxKind.GlobalStatement)
+                    {
+                        node = node.Parent;
+                    }
                     RemoveNode(editor, node, syntaxFacts);
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/RemoveUnusedVariable/CSharpRemoveUnusedVariableCodeFixProvider.cs:66]: same feedback.

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/StatementSyntaxExtensions.cs [hunk 6]
```diff
@@ -29,10 +29,25 @@ public static StatementSyntax GetNextStatement(this StatementSyntax statement)
...
+                {
+                    return false;
+                }
+
+                if (!one.IsParentKind(SyntaxKind.GlobalStatement))
+                {
+                    return false;
+                }
+
+                return one.Parent.Parent == other.Parent.Parent;
```
Comment (CyrusNajmabadi) [inline src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/StatementSyntaxExtensions.cs:49]:
```suggestion
                return one.IsParentKind(SyntaxKind.GlobalStatement) &&
                       one.IsParentKind(SyntaxKind.GlobalStatement) &&
                       one.Parent.Parent == other.Parent.Parent;
```

File: src/Analyzers/CSharp/Tests/ConvertSwitchStatementToExpression/ConvertSwitchStatementToExpressionTests.cs [hunk 7]
```diff
@@ -1890,5 +1890,78 @@ static void GetRef(int[] mem, int addr, int mode)
...
+};
+";
+
+            await VerifyCS.VerifyCodeFixAsync(source, new[] {
+                    // error CS9004: Program using top-level statements must be an executable.
+                    DiagnosticResult.CompilerError("CS9004"),
+                    // error CS8652: The feature 'top-level statements' is currently in Preview and *unsupported*. To use Preview features, use the 'preview' language version.
+                    DiagnosticResult.CompilerError("CS8652").WithSpan(2, 1, 2, 11).WithArguments("top-level statements"),
+                },
+                fixedSource);
```
Comment (sharwell) [inline src/Analyzers/CSharp/Tests/ConvertSwitchStatementToExpression/ConvertSwitchStatementToExpressionTests.cs:1922]: 📝 The test should be run with this feature enabled

```
await new VerifyCS.Test {
  TestCode = source,
  FixedCode = fixedSource,
  LanguageVersion = LanguageVersion.Preview,
}.RunAsync();
```

File: src/Analyzers/CSharp/Tests/UseThrowExpression/UseThrowExpressionTests.cs [hunk 8]
```diff
@@ -541,5 +541,18 @@ public A(T? t)
     }
 }");
         }
+
+        [WorkItem(44454, "https://github.com/dotnet/roslyn/issues/44454")]
```
Comment (sharwell) [inline src/Analyzers/CSharp/Tests/UseThrowExpression/UseThrowExpressionTests.cs:545]: 📝 Another pull request is already submitted to fix this

File: src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/StatementSyntaxExtensions.cs (hunk 6 above)
Comment (sharwell) [reply src/Workspaces/SharedUtilitiesAndExtensions/Compiler/CSharp/Extensions/StatementSyntaxExtensions.cs:49]: 💭 The last condition is redundant. Global statements are always siblings if they are in the same tree.

File: src/Analyzers/CSharp/Tests/UseThrowExpression/UseThrowExpressionTests.cs (hunk 8 above)
Comment (jcouv) [reply src/Analyzers/CSharp/Tests/UseThrowExpression/UseThrowExpressionTests.cs:545]: Awesome. I'll remove this test.

**Pull request merged**